### ISSUE TITLE
Is the NP calculation correct?

### ISSUE THREAD

**Issue by andrewsuzuki** [SauceLLC/sauce4strava#95]
According to "Training and Racing with a Power Meter" 2nd edition, the first step in the NP calculation is "Starting at the beginning of the data and calculating a 30-second rolling average for power".

I noticed in `calcNP` that for the first 30 seconds it effectively includes the averages using `i+1`-second windows (sums seconds 0-1 / 2, then sums seconds 0-2 / 3, then sums seconds 0-3 / 4...), which doesn't seem right.

Instead, I believe it should start immediately at summing seconds 0-29 / 30, then proceed to 1-30/30, etc, all with 30-second windows.

Some non-optimized examples, for 1-second sample rate, that don't handle edge cases.

It's currently doing something equivalent to this:

```js
export function calcNP(data) {
    const windowSize = 30;
    const rollingAverage = [];
    // loop window ending positions
    for (let i = data.length - 1; i >= 0; i--) {
        // sum window
        let sum = 0;
        for (let j = i; j > i - windowSize && j >= 0; j--) {
            sum += data[j];
        }
        rollingAverage.push(sum / (i < windowSize - 1 ? i + 1 : windowSize));
    }
    // calculate NP from rolling average
    return (
        (rollingAverage.reduce((acc, r) => acc + r ** 4, 0) / rolling.length) ** 0.25
    );
}
```

Here's what I think it should be doing:

```js
export function calcNP(data) {
    const windowSize = 30;
    const rollingAverage = [];
    // loop window starting positions
    for (let i = 0; i <= data.length - windowSize; i++) { // CHANGED
        // sum window
        let sum = 0;
        for (let j = i; j < i + windowSize; j++) { // CHANGED
            sum += data[j];
        }
        rollingAverage.push(sum / windowSize); // CHANGED
    }
    // calculate NP from rolling average
    return (
        (rollingAverage.reduce((acc, r) => acc + r ** 4, 0) / rollingAverage.length) ** 0.25
    );
}
```

Playing around with it and comparing it to the current implementation, the difference can be a few watts for shorter rides.

Unless I'm misunderstanding something or Coggan has updated the algorithm since the 2nd edition.

**Comment by mayfield** [comment]
That's interesting.  I actually didn't base my impl on the literature.  I referenced other open source implementations as I recall.  It was many years ago though.  I'll have a look around.  I will say, increasing the O complexity like this will be a bitter pill to swallow, so I hope you're wrong. lol :smile:

**Comment by mayfield** [comment]
Just checked a couple sources and the algo is as you stated; here is the pseudo code description..

![image](https://user-images.githubusercontent.com/139316/157146711-62c8ab24-02ba-4aa8-b246-1f9bc259cde9.png)

It took me a second to see that your inner loops are used to recompute averages of the 30s window but are also achievable with the sum val maintained outside the loop.  As is how the cur impl works and how GC works too (see below).  I'm not sure why your first function is going backwards through the data stream; was that intentional?

Here is GC's impl..

![image](https://user-images.githubusercontent.com/139316/157147543-c06153da-6107-4c28-806c-9187d41951c7.png)

So Sauce is pretty similar to the GC approach with the exception of not zero padding the rolling window.  For example GC's first avg is `[0, 0, ..., 0, 500] / 30` where 500 was the actual first value in the data stream.

Sauce avoids zero padding by monitoring the effective window size as it begins.   Anything over 20 mins it's probably not statistically significant and my anecdotal comparisons with GC come up with same results.  The changes I made were mostly because I got pressured into lowering the NP minimum period from 20 mins (Coggans recommendation) to 5 mins where the zero padding was pretty obvious, especially for selections like in the analysis view.

However a couple things have changed since then.  I now have an optional `leadin` stream arg will essentially seed the window with real data before your actual calcs are provided.  This largely solves over and under weighting the beginning of a selection analysis.

I certainly don't mind diverging from the GC method if it's more true to the original paper but I do need to think more on the alignment effects it will have for viewing selections of data.

**Comment by andrewsuzuki** [comment]
I just wrote those to demonstrate the difference, disregarding time complexity (actually, I almost just posted them in Clojure because that's what I'm currently implementing them in, hah).

Here's a better O(n) version that's about 30% faster than the current impl. Instead of using an intermediate array it maintains a sliding sum of the window. It skips until the first full window's sum is built.

```js
function calcNP(data) {
    const windowSize = 30;
    let summedRollingAveragesToFourthPowers = 0;
    let slidingSum = 0;
    for (let i = 0; i < data.length; i++) {
        slidingSum += data[i];
        if (i < windowSize - 1) {
            // skip until first window sum is built up
            continue;
        }
        const ra = slidingSum / windowSize; // rolling average
        summedRollingAveragesToFourthPowers += ra * ra * ra * ra;
        slidingSum -= data[i - windowSize + 1];
    }
    return (
        (summedRollingAveragesToFourthPowers /
            (data.length - windowSize + 1)) **
        0.25
    );
}
```

I wonder if `leadin` would still be necessary with the above impl?

**Comment by mayfield** [comment]
I've just done a very simple test that think is more or less on point with Coggan's description of prefilling the window...

```patch
diff --git a/src/common/lib.js b/src/common/lib.js
index 107babd..dae1156 100644
--- a/src/common/lib.js
+++ b/src/common/lib.js
@@ -1110,14 +1110,15 @@ sauce.ns('power', function() {
             }
             sum -= rolling[index] || 0;
             rolling[index] = watts;
-            const rollUsedSize = i + 1 + breakPadding;
-            const avg = sum / (rollingSize < rollUsedSize ? rollingSize : rollUsedSize);
-            const qavg = avg * avg * avg * avg;  // About 100 x faster than Math.pow and **
-            total += qavg;
-            if (i < leadin.length) {
-                leadinTotal += qavg;
+            if (i + 1 + breakPadding >= rollingSize) {
+                const avg = sum / rollingSize;
+                const qavg = avg * avg * avg * avg;  // About 100 x faster than Math.pow and **
+                total += qavg;
+                if (i < leadin.length) {
+                    leadinTotal += qavg;
+                }
+                count++;
             }
-            count++;
         }
         count -= leadin.length;
         total -= leadinTotal;
```

**Comment by andrewsuzuki** [comment]
What is Coggan's description of prefilling the window? If the first window is from indices 0 to 29 then isn't it already filled?

**Comment by mayfield** [comment]
Prefilling meaning that before using a `windowAvg^4` value for the NP calc you ensure you have collected the first 30 seconds of data.  The total of windowAvg^4 values is then divided by the length of the input power stream minus 29 (assuming 1 sec data samples).

Assuming 1 second samples with data 
like `[0, 1, 2, 3, ..., 30]` will yield a result of 15.024942348560666 which is the result of this process...

```javascript
sauce.data.avg(data.slice(0, 30)) // 14.5
sauce.data.avg(data.slice(1, 31)) // 15.5
((14.5**4 + 15.5**4) / 2)**0.25 // 15.024942348560666
```

**Comment by mayfield** [comment]
I just copy/pasted your code into a console and it's consistent. 

![image](https://user-images.githubusercontent.com/139316/157543077-d84517af-b899-40b8-a6f9-a3ba1c62afec.png)

**Comment by andrewsuzuki** [comment]
That looks good! The output matches the output of implementation. (Sorry, disregard my deleted comment asking about prefilling, I thought you had done something different)

Then what about the leadin? Is it still necessary if the first 30s is filled?

**Comment by mayfield** [comment]
Fixed in 855dd527fcbbc51c39d050e43f2bcf22b728afbc
